File: 백준/Silver/1991. 트리 순회/트리 순회.py
import sys

N = int(sys.stdin.readline())
tree = {}

for i in range(N):
    root, left, right = sys.stdin.readline().strip().split()
    tree[root] = [left,right]

# 전위 순회
def preorder(node):
    if node != '.':
        print(node, end='')
        preorder(tree[node][0])
        preorder(tree[node][1])


#  중위 순회
def inorder(node):
    # 자식이 있다면
    if node !='.':
        inorder(tree[node][0])
        print(node, end = '')
        inorder(tree[node][1])

# # 후위 순회
def postorder(node):
    if node !='.':
        postorder(tree[node][0])
        postorder(tree[node][1])
        print(node, end='')

preorder('A')
print()
inorder('A')
print()
postorder('A')

File: 백준/Silver/1991. 트리 순회/트리 순회.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const N = parseInt(input.shift());
const graph = new Map();

for (let i = 0; i < N; i++) {
  const [node, left, right] = input[i].trim().split(" ");
  graph.set(node, [left, right]);
}

//console.log(graph);
// 전위 순회 루트>왼>오
function preorder(node, answer) {
  if (node == ".") return "";

  answer.push(node);
  preorder(graph.get(node)[0], answer);
  preorder(graph.get(node)[1], answer);
}

// 중위 순회
function inorder(node, answer) {
  if (node == ".") return "";

  inorder(graph.get(node)[0], answer);
  answer.push(node);
  inorder(graph.get(node)[1], answer);
}

// 후위 순회
function postorder(node, answer) {
  if (node == ".") return "";

  postorder(graph.get(node)[0], answer);
  postorder(graph.get(node)[1], answer);
  answer.push(node);
}

let answerPre = [];
let answerIn = [];
let answerPost = [];

preorder("A", answerPre);
inorder("A", answerIn);
postorder("A", answerPost);

console.log(answerPre.join(""));
console.log(answerIn.join(""));
console.log(answerPost.join(""));
